scan-external-contacts/nodejs/index.js

```diff
@@ -0,0 +1,296 @@
+const maxRateLimitRetryAttempts = 2;
+// - Settings for Export to file (CSV or RAW)
+const fieldSeparator = ',';
+const exportType = 'CSV';
+
+// Global vars
+var writeStream, csvStream;
+var consecutiveRateLimitRetryAttempts = 0;
+
+/*
+------------------------------------------------------------------------------
+* Login, Scan Objects
+------------------------------------------------------------------------------
+*/
+
+//#region Login, Scan
+
+// Login using Client Credentials Grant (returns: true (success), false (failure))
+const login = async () => {
+
+    try {
+        apiClient.setEnvironment(apiClientRegion);
+
+        let creds = await apiClient.loginClientCredentialsGrant(apiClientId, apiClientSecret);
+
+        console.log('Login Success. Token will expire at: ', creds.tokenExpiryTime);
+
+        return true;
+    }
+    catch (error) {
+        console.error('Login error - Error:', JSON.stringify(error, null, 2));
+
+        return false;
+    }
+}
+
+// Perform an External Contacts Scan API request and manage errors (429)
+const scanContacts = async (cursor) => {
+
+    // Defining the default values for the result structure
+    let result = {
+        error: false,
+        errorMsg: "",
+        entities: [],
+        isEmpty: true,
+        isFinished: true,
+        nextCursor: null,
+        startTime: 0
+    };
+
+    try {
+        result.startTime = Date.now();
+
+        let opts = {
+            'limit': maxRecordsPerPage
+        };
+        // No opts.cursor on first scan - cursor == null
+        if (cursor) {
+            opts.cursor = cursor;
+        }
+
+        // External Contacts Scan API request
+        let scanData = await contactsApi.getExternalcontactsScanContacts(opts);
+        // If the request is successful, we analyze the response
+
+        // The response may contain an empty "entities" array in certain conditions (congestion on database, contacts deleted before the request was performed, ...)
+        if (scanData.entities && scanData.entities.length > 0) {
+            result.isEmpty = false;
+            result.entities = scanData.entities;
+        }
+
+        // If the response contains an "after" attribute inside "cursors" (cursors.after), then, there are still some External Contacts (data) to retrieve
+        // When the response does not contain an "after" attribute inside "cursors", it means it is the last page of scan
+        if (scanData.hasOwnProperty("cursors") && scanData.cursors.hasOwnProperty("after")) {
+            result.isFinished = false;
+            result.nextCursor = scanData.cursors.after;
+        }
+
+        // Reset of some global vars - counters for retry
+        consecutiveRateLimitRetryAttempts = 0;
+
+        return result;
+    }
+    catch (error) {
+        console.log('External Contacts Scan error - Error:', JSON.stringify(error, null, 2));
+
+        if (error.status && error.status === 429) {
```

Comment:
The JS SDK has automatic retry logic on 429's.  This would be a great place to show case this.

Reply:
It doesn't currently. DEVENGAGE-271